File: flask_app/app/static/metrics/metrics.js
async function fetchMetrics() {
    const response = await fetch("/metrics-json");
    const data = await response.json();

    // Update bars
    document.getElementById("cpu-bar").style.width = data.cpu_percent + "%";
    document.getElementById("cpu-text").innerText = data.cpu_percent + "%";

    document.getElementById("mem-bar").style.width = data.memory_percent + "%";
    document.getElementById("mem-text").innerText = `${data.memory_used.toFixed(1)} / ${data.memory_total.toFixed(1)} GB (${data.memory_percent}%)`;

    document.getElementById("disk-bar").style.width = data.disk_percent + "%";
    document.getElementById("disk-text").innerText = `${data.disk_used.toFixed(1)} / ${data.disk_total.toFixed(1)} GB (${data.disk_percent}%)`;
}

// Poll every 2 seconds
setInterval(fetchMetrics, 2000);

File: flask_app/app/routes/metrics.py
import logging
logger = logging.getLogger(__name__)

import random
import time

from flask import Blueprint, render_template, session, jsonify, Response
from flask_login import login_required
 

from prometheus_client import Counter, generate_latest

import psutil

metrics_bp = Blueprint("metrics", __name__)

REQUEST_COUNT = Counter("app_requests_total", "Total requests")

@metrics_bp.before_request
def count_requests():
    REQUEST_COUNT.inc()


def process_request(t):
    time.sleep(t)

def get_system_metrics():
    mem = psutil.virtual_memory()
    disk = psutil.disk_usage('/')
    return {
        "cpu_percent": psutil.cpu_percent(interval=0.5),
        "memory_used": mem.used / (1024**3),      # GB
        "memory_total": mem.total / (1024**3),
        "memory_percent": mem.percent,
        "disk_used": disk.used / (1024**3),
        "disk_total": disk.total / (1024**3),
        "disk_percent": disk.percent
    }

@metrics_bp.route("/metrics/prom")
def metrics_count():
    return Response(generate_latest(), mimetype="text/plain")


@metrics_bp.route("/metrics")
@login_required
def metrics_page():        
    metrics = get_system_metrics()
    return render_template("metrics/metrics.html", metrics=metrics)


@metrics_bp.route("/metrics-json")
@login_required
def metrics_json():
    return jsonify(get_system_metrics())
